File: tools/bwapi/bwapi_4_4_abi_fixture.cpp
// Compile with the v4.4.0 SDK include directory and run on both Windows targets.
// It emits the record sizes used by game/src/bwapi/wire.rs.
#include <cstddef>
#include <cstdio>

#include <BWAPI/Client/GameData.h>
#include <BWAPI/Client/GameTable.h>

#define RECORD(T) std::printf("\"" #T "\":%zu,", sizeof(T))
#define FIELD(T, F) std::printf("\"" #T "." #F "\":%zu,", offsetof(T, F))

int main() {
  std::printf("{");
  RECORD(BWAPI::GameInstance);
  RECORD(BWAPI::GameTable);
  RECORD(BWAPI::ForceData);
  RECORD(BWAPI::PlayerData);
  RECORD(BWAPI::UnitData);
  RECORD(BWAPI::BulletData);
  RECORD(BWAPI::RegionData);
  RECORD(BWAPIC::Event);
  RECORD(BWAPIC::Shape);
  RECORD(BWAPIC::Command);
  RECORD(BWAPIC::UnitCommand);
  RECORD(BWAPI::unitFinder);
  RECORD(BWAPI::GameData);
  FIELD(BWAPI::GameData, client_version);
  FIELD(BWAPI::GameData, players);
  FIELD(BWAPI::GameData, units);
  FIELD(BWAPI::GameData, isWalkable);
  FIELD(BWAPI::GameData, events);
  FIELD(BWAPI::GameData, strings);
  FIELD(BWAPI::GameData, commands);
  FIELD(BWAPI::GameData, unitCommands);
  std::printf("\"end\":0}\n");
}

File: tools/bwapi/bwapi_4_4_abi_fixture.rs
//! Emits the Rust ABI anchors consumed by `verify-wire.py`.
//!
//! This is intentionally a standalone `rustc` fixture rather than a Cargo target: it imports
//! the production declarations directly and can be compiled for x86 and x64 independently.

#[path = "../../game/src/bwapi/wire.rs"]
mod wire;

use std::mem::{offset_of, size_of};
use wire::*;

macro_rules! record {
    ($type:ty) => {
        print!("\"{}\":{},", stringify!($type), size_of::<$type>())
    };
}

macro_rules! field {
    ($type:ty, $field:ident) => {
        print!(
            "\"{}.{}\":{},",
            stringify!($type),
            stringify!($field),
            offset_of!($type, $field),
        )
    };
}

fn main() {
    print!("{{");
    record!(GameInstance);
    record!(GameTable);
    record!(ForceData);
    record!(PlayerData);
    record!(UnitData);
    record!(BulletData);
    record!(RegionData);
    record!(Event);
    record!(Shape);
    record!(Command);
    record!(UnitCommand);
    record!(UnitFinder);
    record!(GameData);
    field!(GameData, client_version);
    field!(GameData, players);
    field!(GameData, units);
    field!(GameData, is_walkable);
    field!(GameData, events);
    field!(GameData, strings);
    field!(GameData, commands);
    field!(GameData, unit_commands);
    println!("\"end\":0}}");
}
